### DISCUSSION TITLE
Do you plan to port ThreadLocalSingleton and ThreadSafeSingleton?

### DISCUSSION THREAD

**Discussion by nhtgl** [modern-python/that-depends#105] [Q&A] [answered]
As in the title. I'm talking about this functionality https://python-dependency-injector.ets-labs.org/providers/singleton.html#using-singleton-with-multiple-threads. This saved our application. 

Maybe your implementation of Singleton is already thread safe. I'd be glad to learn more on those topics from you!

**Comment by lesnik512** [comment]
@macieyng Hi and thank you for sharing this! Our Singleton is not thread safe yet, but we'll look into this

**Reply by lesnik512** [reply]
I think, that ThreadSafeSingleton is must have and here is the issue for it https://github.com/modern-python/that-depends/issues/106
For ThreadLocalSingleton, I am in doubt if it's really needed

**Reply by nhtgl** [reply]
Correct me ig I'm wrong, but:
- I think that ThreadSafeSingleton is a good use case when you have a main thread creating an instance and subthreads use this singleton.
- ThreadLocalSingleton is good when each of the subthreads should create their own singleton. This what was used in our case.

I understand that they serve different purposes. Would implemented and released Singleton work in both cases?

**Reply by lesnik512** [reply]
Singleton here is more like ThreadSafeSingleton

for ThreadLocalSingleton I created another issue https://github.com/modern-python/that-depends/issues/126

**Comment by lesnik512** [comment, marked as answer]
Here it is. Now singletons and resources are thread-safe in sync resolving:

https://github.com/modern-python/that-depends/releases/tag/1.25.0
